google-oauth-client/src/main/java/com/google/api/client/auth/oauth2/AuthorizationCodeFlow.java

```diff
@@ -412,6 +440,71 @@ public final Clock getClock() {
     void onCredentialCreated(Credential credential, TokenResponse tokenResponse) throws IOException;
   }
 
+  /**
+   * An implementation of <a href="https://tools.ietf.org/html/rfc7636">Proof Key for Code Exchange</a>
+   * which, according to the <a href="https://tools.ietf.org/html/rfc8252#section-6">OAuth 2.0 for Native Apps RFC</a>,
+   * is mandatory for public native apps.
+   */
+  private static class PKCE {
+    private final String verifier;
+    private final String challenge;
+    private String challengeMethod;
+
+    public PKCE() {
+      verifier = generateVerifier();
+      Challenge c = new Challenge(verifier);
+      challenge = c.challenge;
+      challengeMethod = c.method;
+    }
+
+    private String generateVerifier() {
+      SecureRandom sr = new SecureRandom();
+      byte[] code = new byte[32];
+      sr.nextBytes(code);
+      return Base64.encodeBase64URLSafeString(code);
+    }
+
+    public String getVerifier() {
+      return verifier;
+    }
+
+    public String getChallenge() {
+      return challenge;
+    }
+
+    public String getChallengeMethod() {
+      return challengeMethod;
+    }
+
+    /**
+     * An abstraction to create the PKCE code verifier. It will use the S256 method but
+     * will fall back to using the 'plain' method in the unlikely case that the SHA-256
+     * MessageDigest algorithm implementation can't be loaded.
+     */
+    private static class Challenge {
```

Comment:
It seems like this functionality could be rolled into the PKCE class as a static method since it's only used during initialization in PKCE

Reply:
I'll create a method from this. It can't easily be static though unless I return a "Pair<String, String>" or a String array of size 2 to represent the challange and method. But a private method that initialises the challenge/method values is plenty good here I think.